dokka-runners/dokka-gradle-plugin/src/main/kotlin/internal/PluginFeaturesService.kt

```diff
@@ -193,43 +196,81 @@ internal abstract class PluginFeaturesService : BuildService<PluginFeaturesServi
+                    parameters(configureServiceParams)
+                    // This service was successfully registered, so it is considered 'primary'.
+                    parameters.primaryService.set(true)
+                }.get()
+            } catch (ex: ClassCastException) {
+                try {
+                    // Recover from Gradle bug: re-register the service, but don't mark it as 'primary'.
+                    project.gradle.sharedServices.registerIfAbsent(
+                        PluginFeaturesService::class,
+                        classLoaderScoped = true,
+                    ) {
+                        parameters(configureServiceParams)
+                        parameters.primaryService.set(false)
+                    }.get()
+                } catch (ex: ClassCastException) {
+                    throw GradleException(
+                        "Failed to register BuildService. Please report this problem https://github.com/gradle/gradle/issues/17559",
+                        ex
                     )
                 }
+            }
+        }
 
-                return try {
-                    gradle.sharedServices.registerIfAbsent(PluginFeaturesService::class) {
-                        parameters(setFlags)
-                        // This service was successfully registered, so it is considered 'primary'.
-                        parameters.primaryService.set(true)
-                    }.get()
-                } catch (ex: ClassCastException) {
-                    try {
-                        // Recover from Gradle bug: re-register the service, but don't mark it as 'primary'.
-                        gradle.sharedServices.registerIfAbsent(
-                            PluginFeaturesService::class,
-                            classLoaderScoped = true,
-                        ) {
-                            parameters(setFlags)
-                            parameters.primaryService.set(false)
-                        }.get()
-                    } catch (ex: ClassCastException) {
-                        throw GradleException(
-                            "Failed to register BuildService. Please report this problem https://github.com/gradle/gradle/issues/17559",
-                            ex
-                        )
+        private fun Project.serviceParamsConfiguration(): Action<Params> = Action {
+            v2PluginEnabled.set(getFlag(V2_PLUGIN_ENABLED_FLAG))
+            v2PluginNoWarn.set(getFlag(V2_PLUGIN_NO_WARN_FLAG_PRETTY).orElse(getFlag(V2_PLUGIN_NO_WARN_FLAG)))
+            v2PluginMigrationHelpersEnabled.set(getFlag(V2_PLUGIN_MIGRATION_HELPERS_FLAG))
+            k2AnalysisEnabled.set(getFlag(K2_ANALYSIS_ENABLED_FLAG))
+            k2AnalysisNoWarn.set(
+                getFlag(K2_ANALYSIS_NO_WARN_FLAG_PRETTY)
+                    .orElse(getFlag(K2_ANALYSIS_NO_WARN_FLAG))
+            )
+
+            try {
```

Comment:
It would be nice to extract this into separate function with a high-level description on why we are doing this